File: external_wop/wop/estimation.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable

import numpy as np


@dataclass(frozen=True)
class TrajectoryResult:
    value: float
    steps: int
    status: str  # hit_face | timeout | escaped


@dataclass(frozen=True)
class EstimateResult:
    J: float
    S2: float
    eps: float
    n_total: int
    n_truncated: int
    mean_steps: float


class _EstimateAccumulator:
    def __init__(self) -> None:
        self.j_sum = 0.0
        self.s2_raw_sum = 0.0
        self.steps_sum = 0
        self.n_timeout = 0
        self.n_escaped = 0

    def add(self, tr: TrajectoryResult) -> None:
        ksi = float(tr.value)
        self.j_sum += ksi
        self.s2_raw_sum += ksi * ksi
        self.steps_sum += int(tr.steps)

        if tr.status == "timeout":
            self.n_timeout += 1
        elif tr.status == "escaped":
            self.n_escaped += 1

    def finalize(self, n_total: int) -> EstimateResult:
        if n_total <= 0:
            raise ValueError("n_total must be positive.")

        j = self.j_sum / n_total
        s2 = self.s2_raw_sum / n_total - j * j
        s2 = max(float(s2), 0.0)
        eps = 3.0 * float(np.sqrt(s2 / n_total))
        n_truncated = self.n_timeout + self.n_escaped

        return EstimateResult(
            J=j,
            S2=s2,
            eps=eps,
            n_total=n_total,
            n_truncated=n_truncated,
            mean_steps=float(self.steps_sum / n_total),
        )


def estimate_from_trajectories(
    n_paths: int,
    trace_once: Callable[[], TrajectoryResult],
) -> EstimateResult:
    """Generic Monte Carlo aggregation for trajectory-based estimators."""
    if n_paths <= 0:
        raise ValueError("n_paths must be positive.")

    acc = _EstimateAccumulator()
    add = acc.add
    for _ in range(n_paths):
        add(trace_once())
    return acc.finalize(n_paths)

File: external_wop_cpp/src/estimation/estimation.cpp
#include "wop/estimation/estimation.hpp"

#include <cmath>
#include <cstdint>
#include <stdexcept>

namespace wop::estimation {

namespace {

class EstimateAccumulator {
public:
    void add(const TrajectoryResult& tr) {
        ++count_;
        const double delta = tr.value - mean_;
        mean_ += delta / count_;
        m2_ += delta * (tr.value - mean_);
        steps_sum_ += static_cast<std::int64_t>(tr.steps);

        if (tr.status == TrajectoryStatus::Timeout) {
            ++n_timeout_;
        } else if (tr.status == TrajectoryStatus::Escaped) {
            ++n_escaped_;
        }
    }

    EstimateResult finalize(int n_total) const {
        if (n_total <= 0) {
            throw std::invalid_argument("n_total must be positive.");
        }

        const double j = mean_;
        const double s2 = std::max(m2_ / static_cast<double>(n_total), 0.0);
        const double eps = 3.0 * std::sqrt(s2 / static_cast<double>(n_total));
        const int n_truncated = n_timeout_ + n_escaped_;

        return EstimateResult{
            j,
            s2,
            eps,
            n_total,
            n_truncated,
            static_cast<double>(steps_sum_) / static_cast<double>(n_total),
        };
    }

private:
    int count_ = 0;
    double mean_ = 0.0;
    double m2_ = 0.0;
    std::int64_t steps_sum_ = 0;
    int n_timeout_ = 0;
    int n_escaped_ = 0;
};

}  // namespace

EstimateResult estimate_from_trajectories(int n_paths, const std::function<TrajectoryResult()>& trace_once) {
    if (n_paths <= 0) {
        throw std::invalid_argument("n_paths must be positive.");
    }

    EstimateAccumulator acc;
    for (int i = 0; i < n_paths; ++i) {
        acc.add(trace_once());
    }
    return acc.finalize(n_paths);
}

}  // namespace wop::estimation
